Efficient "repo stat" (DiskUsage) and "--size-only" flag

hsanjuan opened the pull request:
This makes use of the PersistentDatastore DiskUsage method to obtain the Repo's storage usage (GetStorageUsage()).
    
Additionally, the --size-only flag has been added to the "ipfs repo stat" command. This avoids counting the number of objects in the repository and returns faster.

Overseeds #4550 

I had to make some gx magic to fix dependencies and be able to use latest go-datastore. go-libp2p-kad-dht depends on it and afaik the changes there cannot be integrated still etc. I'm happy to leave this hanging until this is sorted out. Thus, tagging this as blocked.

hsanjuan commented:
@whyrusleeping is that the case now?

whyrusleeping commented:
@hsanjuan yeah, #5007 merged

hsanjuan commented:
Now rebased on master. Thanks @Stebalien for bubbling up deps on some other commit.

Stebalien commented on core/commands/repo.go:
```diff
@@ -203,9 +215,14 @@ Version         string The repo version.
+				return nil
+			}
+
+			fmt.Fprintf(wtr, "NumObjects:\t%d\n", stat.NumObjects)
```
This is technically a breaking change but I doubt anyone is relying on this order.

hsanjuan replied:
I hope not. I can turn it around though and leave NumObjects as first line.

hsanjuan replied:
I'll just change it. It's marginally better to not change the "API".

Stebalien commented on core/corerepo/stat.go:
```diff
@@ -45,11 +46,29 @@ func RepoStat(n *core.IpfsNode, ctx context.Context) (*Stat, error) {
+}
+
+// RepoSize returns a *Stat object with the RepoSize and StorageMax fields set.
+func RepoSize(ctx context.Context, n *core.IpfsNode) (*Stat, error) {
```
This feels a bit funky... what about:

```go
type Size struct {
	RepoSize   uint64
	StorageMax uint64
}

type Stat struct {
	StorageSize
	NumObjects  uint64
	RepoPath    string
	Version     string
}
```

And then have `RepoSize` return `Size` (and, really, these functions should return be returning these by value). This would also avoid sending zero values for the missing fields.

Thoughts?

hsanjuan replied:
yes this is better

hsanjuan commented:
@Stebalien I have addressed the comments. However, since the command output only supports a single type, `SizeStat` needs to eventually become `Stat`. The final result is 0-valued json fields when doing a request with --size-only. (this is ok for me). If we wanted to return json without those fields, we would need a second command (`ipfs repo size`).

Stebalien commented on core/commands/repo.go:
```diff
@@ -150,65 +150,85 @@ var repoStatCmd = &cmds.Command{
 		cmds.Text: cmds.MakeEncoder(func(req *cmds.Request, w io.Writer, v interface{}) error {
 			stat, ok := v.(*corerepo.Stat)
 			if !ok {
+				fmt.Println("adios")
```
Debug print statement.

hsanjuan replied:
damn

hsanjuan replied:
fixed

hsanjuan commented:
@whyrusleeping can we merge this please?

Stebalien commented:
@hsanjuan once we do the release which should happen any day now.
